src/zmq/zmqpublishnotifier.cpp

```diff
@@ -171,3 +177,41 @@ bool CZMQPublishRawTransactionNotifier::NotifyTransaction(const CTransaction &tr
     int rc = zmq_send_multipart(psocket, "rawtx", 5, &(*ss.begin()), ss.size(), 0);
     return rc == 0;
 }
+
+CZMQPublishMempoolNotifier::CZMQPublishMempoolNotifier(CTxMemPool *mempoolIn):
+    mempool(mempoolIn)
+{
+    mempool->NotifyEntryAdded.connect(boost::bind(&CZMQPublishMempoolNotifier::NotifyEntryAdded, this, _1));
+    mempool->NotifyEntryRemoved.connect(boost::bind(&CZMQPublishMempoolNotifier::NotifyEntryRemoved, this, _1, _2));
+}
+
+CZMQPublishMempoolNotifier::~CZMQPublishMempoolNotifier()
+{
+    mempool->NotifyEntryAdded.disconnect(boost::bind(&CZMQPublishMempoolNotifier::NotifyEntryAdded, this, _1));
+    mempool->NotifyEntryRemoved.disconnect(boost::bind(&CZMQPublishMempoolNotifier::NotifyEntryRemoved, this, _1, _2));
+}
+
+void CZMQPublishMempoolNotifier::NotifyEntryAdded(const CTxMemPoolEntry &entry)
+{
+    LogPrint("zmq", "zmq: mempool entry added: %s fee=%i size=%i\n", entry.GetTx().GetHash().ToString(),
+            entry.GetFee(), entry.GetTxSize());
+    uint8_t data[32 + 8 + 4];
+    WriteHash(&data[0], entry.GetTx().GetHash());
+    WriteLE64(&data[32], entry.GetFee());
+    WriteLE32(&data[40], entry.GetTxSize());
+    int rc = zmq_send_multipart(psocket, "mempooladded", 12, data, sizeof(data), 0);
+    if (rc != 0)
+        LogPrintf("%s: WARNING: zmq send failed with code %i\n", __func__, rc);
+}
+
+void CZMQPublishMempoolNotifier::NotifyEntryRemoved(const CTxMemPoolEntry &entry, MemPoolRemovalReason reason)
+{
+    LogPrint("zmq", "zmq: mempool entry removed: %s, reason %i\n", entry.GetTx().GetHash().ToString(), (int)reason);
+    uint8_t data[32 + 1];
+    WriteHash(&data[0], entry.GetTx().GetHash());
+    data[32] = reason;
```

Comment:
I would prefer it if the value of the (internal) MemPoolRemovalReason did not leak into the ZMQ protocol, though that can be fixed later by introducing a new enum in the ZMQ definition.